File: typescript/Behavioral/17-Memento/Caretaker/JobCardHistory.ts
import { JobCardMemento } from "../Memento/JobCardMemento";

export class JobCardHistory {
  private history: JobCardMemento[] = [];

  push(memento: JobCardMemento): void {
    this.history.push(memento);
    console.log(`  [History] Saved: ${memento}`);
  }

  pop(): JobCardMemento | undefined {
    if (this.history.length === 0) {
      console.log("  [History] Nothing to restore.");
      return undefined;
    }

    const memento = this.history.pop()!;
    console.log(`  [History] Restoring: ${memento}`);
    return memento;
  }

  printAll(): void {
    console.log(`  [History] ${this.history.length} snapshot(s):`);
    for (const m of [...this.history].reverse()) console.log(`    ${m}`);
  }
}

File: csharp/Behavioral/17-Memento/Caretaker/JobCardHistory.cs
public class JobCardHistory
{
    private readonly Stack<JobCard.JobCardMemento> _history = new();

    public void Push(JobCard.JobCardMemento memento)
    {
        _history.Push(memento);
        Console.WriteLine($"  [History] Saved: {memento}");
    }

    public JobCard.JobCardMemento? Pop()
    {
        if (_history.Count == 0)
        {
            Console.WriteLine("  [History] Nothing to restore.");
            return null;
        }

        var memento = _history.Pop();
        Console.WriteLine($"  [History] Restoring: {memento}");
        return memento;
    }

    public void PrintAll()
    {
        Console.WriteLine($"  [History] {_history.Count} snapshot(s):");
        foreach (var m in _history)
            Console.WriteLine($"    {m}");
    }
}
